### PR TITLE
make the empty parens after level constructors optional

### PR DESCRIPTION
**Pull request by ericniebler**
For example, what before needed to be written as:

```cpp
cudax::make_hierarchy(cudax::block_dims<block_size>(), cudax::grid_dims<grid_size>())
```

can now be written as:

```cpp
cudax::make_hierarchy(cudax::block_dims<block_size>, cudax::grid_dims<grid_size>)
```

This syntax cannot be made to work with `operator&`. For example, this will not compile:

```cpp
cudax::block_dims<block_size> & cudax::grid_dims<grid_size> // compile error
```

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cudax/include/cuda/experimental/__hierarchy/hierarchy_dimensions.cuh [hunk 1]
```diff
@@ -103,69 +125,76 @@ struct get_level_helper
...
+  _CCCL_NODISCARD _CUDAX_TRIVIAL_API static constexpr auto
+  __apply_reverse(const Levels& ls, ::cuda::std::index_sequence<_Ids...>) noexcept
   {
-    return hierarchy_dimensions_fragment(LUnit{}, ::cuda::std::forward<L1>(l1), ::cuda::std::forward<Levels>(ls)...);
+    return __make_hierarchy_fragment<LUnit, true>()(::cuda::std::get<_Ids>(ls)...);
   }
-  else
+
+  template <typename L1, typename... Levels>
+  _CCCL_NODISCARD _CUDAX_API constexpr auto operator()(const L1& l1, const Levels&... ls) const noexcept
```
Comment (pciolkosz) [inline cudax/include/cuda/experimental/__hierarchy/hierarchy_dimensions.cuh:166]: I think we can now drop `l1`? It was needed previously for the `can_stack`, but now I think it's always used along with the ls pack.

File: cudax/include/cuda/experimental/__hierarchy/hierarchy_dimensions.cuh [hunk 2]
```diff
@@ -767,45 +792,44 @@ constexpr auto make_hierarchy_fragment(L1&& l1, Levels&&... ls) noexcept
...
+  return detail::__make_hierarchy_fragment<thread_level>()(detail::__as_level(l1), detail::__as_level(ls)...);
 }
 
-// We can consider removing the operator, but its convinient for in-line construction
+// We can consider removing the operator&, but its convenient for in-line construction
 // TODO accept forwarding references
 template <typename LUnit, typename L1, typename... Levels>
-_CCCL_HOST_DEVICE constexpr auto
-operator&(const hierarchy_dimensions_fragment<LUnit, Levels...>& ls, const L1& l1) noexcept
+_CUDAX_API constexpr auto operator&(const hierarchy_dimensions_fragment<LUnit, Levels...>& ls, L1 l1) noexcept
```
Comment (pciolkosz) [inline cudax/include/cuda/experimental/__hierarchy/hierarchy_dimensions.cuh:803]: Nit: With the changes I noticed how bad L1/l1 name is here, its a copy paste from other places probably. Maybe `new_level` or similar would be better here?

**Review by pciolkosz — APPROVED** [review]
Thanks for looking into it

File: cudax/include/cuda/experimental/__hierarchy/hierarchy_dimensions.cuh (hunk 1 above)
Comment (ericniebler) [reply cudax/include/cuda/experimental/__hierarchy/hierarchy_dimensions.cuh]: `__can_stack` wants to compare the shifted list of levels. we need to separate out `L1` here so we can easily shift the levels.

Comment (pciolkosz) [reply cudax/include/cuda/experimental/__hierarchy/hierarchy_dimensions.cuh]: Hmm, its not as easy as I thought, we would also need to change `__can_stack` to not be an alias, not sure if its worth it.
Feel free to revert my change here

**Pull request merged**